src/librustc_mir/borrow_check/nll/region_infer/mod.rs

```diff
@@ -450,21 +438,35 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         }
     }
 
+    pub(crate) fn compute_causal_info(&self, mir: &Mir<'tcx>) -> RegionCausalInfo {
+        self.compute_causal_info2(mir, true)
+    }
+
     /// Propagate the region constraints: this will grow the values
     /// for each region variable until all the constraints are
     /// satisfied. Note that some values may grow **too** large to be
     /// feasible, but we check this later.
     fn propagate_constraints(&mut self, mir: &Mir<'tcx>) {
-        debug!("propagate_constraints()");
-        debug!("propagate_constraints: constraints={:#?}", {
+        let causal_info = self.compute_causal_info2(mir, false);
+        self.inferred_values = Some(causal_info.inferred_values);
+    }
+
+    pub(crate) fn compute_causal_info2(&self, mir: &Mir<'tcx>, track_causes: bool) -> RegionCausalInfo {
+        debug!("compute_causal_info2()");
+        debug!("compute_causal_info2: constraints={:#?}", {
             let mut constraints: Vec<_> = self.constraints.iter().collect();
             constraints.sort();
             constraints
         });
 
         // The initial values for each region are derived from the liveness
         // constraints we have accumulated.
-        let mut inferred_values = self.liveness_constraints.clone();
+        let mut inferred_values =
```

Comment:
Rather than pass a flag, I personally would have this take as argument the initial `RegionValues`, so that the caller clones either with causes or without. I just hate boolean flags. If we are going to have a flag, I'd prefer it to be `struct TrackCauses(bool)`, so that when we see a call to it, it looks like `self.compute_causal_info(..., TrackCauses(true))` and hence we know what that `true/false` means.

Reply:
Had a thought. First off, I see an argument for keeping the flag (though I prefer the struct), because -- as I wrote on gitter -- it shows that it only makes sense for `compute_causal_info` to start with a fresh clone of the liveness data versus some caller that might clone the liveness data, manipulate it in some way, and then invoke `compute_causal_info`. But if we are going to have the flag, then let's remove `clone_tracking_causal_info` thing and pass the flag to `RegionValues`:

```
fn clone(&self, track_causes: TrackCauses) -> RegionValues { ... }
```

Maybe pick another name than `clone` (`duplicate`?)